User Assignment to Service App Causes Panic

An admin assigned themselves to an OAuth application of type service in the user screen of the Okta admin site. After this happened, terraform plans started crashing on the `okta_app_oauth` resource. Removing the user assignment resolved the issue.

<!--- Please keep this note for the community --->

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

0.14.2

### Affected Resource(s)

okta_app_oauth

### Debug Output

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xedcdc6]

goroutine 32 [running]:
github.com/oktadeveloper/terraform-provider-okta/okta.syncGroupsAndUsers(0x13c1fa0, 0xc00007b440, 0xc0002cdc40, 0x14, 0xc0005c4f00, 0x10797a0, 0xc0005c2510, 0x0, 0x0)
	github.com/oktadeveloper/terraform-provider-okta/okta/app.go:433 +0x406
github.com/oktadeveloper/terraform-provider-okta/okta.resourceAppOAuthRead(0x13c1fa0, 0xc00007b440, 0xc0005c4f00, 0x10797a0, 0xc0005c2510, 0xc000429f40, 0xc0004c38f0, 0x40e0f8)
	github.com/oktadeveloper/terraform-provider-okta/okta/resource_okta_app_oauth.go:378 +0xa26
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc00014a900, 0x13c1f20, 0xc0005fc500, 0xc0005c4f00, 0x10797a0, 0xc0005c2510, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.0/helper/schema/resource.go:297 +0x1ea
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00014a900, 0x13c1f20, 0xc0005fc500, 0xc000162d90, 0x10797a0, 0xc0005c2510, 0xc00000ecc8, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.0/helper/schema/resource.go:564 +0x1cb
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc00018c140, 0x13c1f20, 0xc0005fc500, 0xc0005fc540, 0xc0005fc500, 0x1163e20, 0x11b3220)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.0/helper/schema/grpc_provider.go:575 +0x43b
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadResource(0xc000245c20, 0x13c1f20, 0xc0005fc500, 0xc00007aea0, 0xc000245c20, 0xc0003ff080, 0xc000261ba0)
	github.com/hashicorp/terraform-plugin-go@v0.1.0/tfprotov5/server/server.go:297 +0x105
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler(0x11b3220, 0xc000245c20, 0x13c1fe0, 0xc0003ff080, 0xc00007ae40, 0x0, 0x13c1fe0, 0xc0003ff080, 0xc000424000, 0x25a)
	github.com/hashicorp/terraform-plugin-go@v0.1.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001b8c40, 0x13cd5c0, 0xc0000c5c80, 0xc0005b0600, 0xc000180ae0, 0x1a1fef0, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1194 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc0001b8c40, 0x13cd5c0, 0xc0000c5c80, 0xc0005b0600, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1517 +0xd05
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0000a8560, 0xc0001b8c40, 0x13cd5c0, 0xc0000c5c80, 0xc0005b0600)
	google.golang.org/grpc@v1.32.0/server.go:859 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
```

### Panic Output

Sorry, don't have the output from crash.log

### Expected Behavior

The plan succeeds

### Actual Behavior

The provider crashes :'(

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. Write some tf for an oauth app of type service
2. Run `terraform plan/apply` to create the app in Okta
3. In the admin UI, assign a user to the app (this has to be done from the user screen)
4. Run `terraform plan` and it should crash

### Important Factoids

Seems like this is a potential problem: https://github.com/oktadeveloper/terraform-provider-okta/blob/c79db0a92f791a014b782c705a302a0fb4d555f5/okta/app.go#L434

The call to get the app user assignments returns a user with a `null` value for `credentials`:

```json
{
  "id": "12345",
  "externalId": null,
  "created": "2021-01-20T16:50:29.000Z",
  "lastUpdated": "2021-01-20T16:50:29.000Z",
  "scope": "USER",
  "status": "ACTIVE",
  "statusChanged": "2021-01-20T16:50:29.000Z",
  "passwordChanged": null,
  "syncState": "DISABLED",
  "credentials": null,
  ...
}
```

--- a/examples/okta_app_oauth/datasource.tf
+++ b/examples/okta_app_oauth/datasource.tf
@@ -1,0 +1,19 @@
+resource "okta_app_oauth" "test" {
+  label                      = "testAcc_replace_with_uuid"
+  type                       = "web"
+  grant_types                = ["authorization_code"]
+  redirect_uris              = ["http://d.com/"]
+  response_types             = ["code"]
+  client_basic_secret        = "something_from_somewhere"
+  client_id                  = "something_from_somewhere"
+  token_endpoint_auth_method = "client_secret_basic"
+  consent_method             = "TRUSTED"
+}
+
+data "okta_app_oauth" "test" {
+  id = okta_app_oauth.test.id
+}
+
+data "okta_app_oauth" "test_label" {
+  label = okta_app_oauth.test.label
+}

--- a/okta/app.go
+++ b/okta/app.go
@@ -429,9 +429,13 @@ func syncGroupsAndUsers(ctx context.Context, id string, d *schema.ResourceData, 
 
 	for _, user := range userList {
 		if user.Scope == userScope {
+			var un string
+			if user.Credentials != nil {
+				un = user.Credentials.UserName
+			}
 			flattenedUserList = append(flattenedUserList, map[string]interface{}{
 				"id":       user.Id,
-				"username": user.Credentials.UserName,
+				"username": un,
 			})
 		}
 	}

--- a/okta/app.go
+++ b/okta/app.go
@@ -326,6 +326,18 @@ func containsAppUser(userList []*okta.AppUser, id string) bool {
 	return false
 }
 
+func shouldUpdateUser(userList []*okta.AppUser, id, username string) bool {
+	for _, user := range userList {
+		if user.Id == id &&
+			user.Scope == userScope &&
+			user.Credentials != nil &&
+			user.Credentials.UserName != username {
+			return true
+		}
+	}
+	return false
+}
+
 // Handles the assigning of groups and users to Applications. Does so asynchronously.
 func handleAppGroupsAndUsers(ctx context.Context, id string, d *schema.ResourceData, m interface{}) error {
 	var wg sync.WaitGroup
@@ -353,17 +365,14 @@ func handleAppUsers(ctx context.Context, id string, d *schema.ResourceData, clie
 	if set, ok := d.GetOk("users"); ok {
 		users = set.(*schema.Set).List()
 		userIDList = make([]string, len(users))
-
 		for i, user := range users {
 			userProfile := user.(map[string]interface{})
 			uID := userProfile["id"].(string)
+			username := userProfile["username"].(string)
 			userIDList[i] = uID
-
+			// Not required
+			password, _ := userProfile["password"].(string)
 			if !containsAppUser(existingUsers, uID) {
-				username := userProfile["username"].(string)
-				// Not required
-				password, _ := userProfile["password"].(string)
-
 				asyncActionList = append(asyncActionList, func() error {
 					_, _, err := client.Application.AssignUserToApplication(ctx, id, okta.AppUser{
 						Id: uID,
@@ -374,7 +383,19 @@ func handleAppUsers(ctx context.Context, id string, d *schema.ResourceData, clie
 							},
 						},
 					})
-
+					return err
+				})
+			} else if shouldUpdateUser(existingUsers, uID, username) {
+				asyncActionList = append(asyncActionList, func() error {
+					_, _, err := client.Application.UpdateApplicationUser(ctx, id, uID, okta.AppUser{
+						Id: uID,
+						Credentials: &okta.AppUserCredentials{
+							UserName: username,
+							Password: &okta.AppUserPasswordCredential{
+								Value: password,
+							},
+						},
+					})
 					return err
 				})
 			}
@@ -429,13 +450,18 @@ func syncGroupsAndUsers(ctx context.Context, id string, d *schema.ResourceData, 
 
 	for _, user := range userList {
 		if user.Scope == userScope {
-			var un string
+			var un, up string
 			if user.Credentials != nil {
 				un = user.Credentials.UserName
+				if user.Credentials.Password != nil {
+					up = user.Credentials.Password.Value
+				}
 			}
 			flattenedUserList = append(flattenedUserList, map[string]interface{}{
 				"id":       user.Id,
 				"username": un,
+				"scope":    user.Scope,
+				"password": up,
 			})
 		}
 	}

--- a/okta/data_source_okta_app_oauth.go
+++ b/okta/data_source_okta_app_oauth.go
@@ -1,0 +1,187 @@
+package okta
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/okta/okta-sdk-golang/v2/okta"
+	"github.com/okta/okta-sdk-golang/v2/okta/query"
+)
+
+func dataSourceAppOauth() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceAppOauthRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"label", "label_prefix"},
+			},
+			"label": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"id", "label_prefix"},
+			},
+			"label_prefix": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"id", "label"},
+			},
+			"active_only": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "Search only ACTIVE applications.",
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"auto_submit_toolbar": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Display auto submit toolbar",
+			},
+			"hide_ios": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Do not display application icon on mobile app",
+			},
+			"hide_web": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Do not display application icon to users",
+			},
+			"grant_types": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+				Description: "List of OAuth 2.0 grant types",
+			},
+			"response_types": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+				Description: "List of OAuth 2.0 response type strings.",
+			},
+			"redirect_uris": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+				Description: "List of URIs for use in the redirect-based flow.",
+			},
+			"post_logout_redirect_uris": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+				Description: "List of URIs for redirection after logout",
+			},
+			"logo_uri": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URI that references a logo for the client.",
+			},
+			"login_uri": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URI that initiates login.",
+			},
+			"client_uri": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URI to a web page providing information about the client.",
+			},
+			"client_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "OAuth client ID",
+			},
+			"policy_uri": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "URI to web page providing client policy document.",
+			},
+		},
+	}
+}
+
+func dataSourceAppOauthRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	filters, err := getAppFilters(d)
+	if err != nil {
+		return diag.Errorf("invalid OAuth app filters: %v", err)
+	}
+	var app *okta.OpenIdConnectApplication
+	if filters.ID != "" {
+		respApp, _, err := getOktaClientFromMetadata(m).Application.GetApplication(ctx, filters.ID, okta.NewOpenIdConnectApplication(), nil)
+		if err != nil {
+			return diag.Errorf("failed get app by ID: %v", err)
+		}
+		app = respApp.(*okta.OpenIdConnectApplication)
+	} else {
+		re := getOktaClientFromMetadata(m).GetRequestExecutor()
+		qp := &query.Params{Limit: 1, Filter: filters.Status, Q: filters.getQ()}
+		req, err := re.NewRequest("GET", fmt.Sprintf("/api/v1/apps%s", qp.String()), nil)
+		if err != nil {
+			return diag.Errorf("failed to list OAuth apps: %v", err)
+		}
+		var appList []*okta.OpenIdConnectApplication
+		_, err = re.Do(ctx, req, &appList)
+		if err != nil {
+			return diag.Errorf("failed to list OAuth apps: %v", err)
+		}
+		if len(appList) < 1 {
+			return diag.Errorf("no OAuth application found with provided filter: %s", filters)
+		}
+		if filters.Label != "" && appList[0].Label != filters.Label {
+			return diag.Errorf("no OAuth application found with the provided label: %s", filters.Label)
+		} else {
+			logger(m).Info("found multiple OAuth applications with the criteria supplied, using the first one, sorted by creation date")
+			app = appList[0]
+		}
+	}
+	d.SetId(app.Id)
+	_ = d.Set("label", app.Label)
+	_ = d.Set("name", app.Name)
+	_ = d.Set("status", app.Status)
+	_ = d.Set("type", app.Settings.OauthClient.ApplicationType)
+	_ = d.Set("auto_submit_toolbar", app.Visibility.AutoSubmitToolbar)
+	_ = d.Set("hide_ios", app.Visibility.Hide.IOS)
+	_ = d.Set("hide_web", app.Visibility.Hide.Web)
+	_ = d.Set("client_uri", app.Settings.OauthClient.ClientUri)
+	_ = d.Set("logo_uri", app.Settings.OauthClient.LogoUri)
+	_ = d.Set("login_uri", app.Settings.OauthClient.InitiateLoginUri)
+	_ = d.Set("client_id", app.Credentials.OauthClient.ClientId)
+	_ = d.Set("policy_uri", app.Settings.OauthClient.PolicyUri)
+	respTypes := make([]string, len(app.Settings.OauthClient.ResponseTypes))
+	for i := range app.Settings.OauthClient.ResponseTypes {
+		respTypes[i] = string(*app.Settings.OauthClient.ResponseTypes[i])
+	}
+	grantTypes := make([]string, len(app.Settings.OauthClient.GrantTypes))
+	for i := range app.Settings.OauthClient.GrantTypes {
+		grantTypes[i] = string(*app.Settings.OauthClient.GrantTypes[i])
+	}
+	aggMap := map[string]interface{}{
+		"redirect_uris":             convertStringSetToInterface(app.Settings.OauthClient.RedirectUris),
+		"response_types":            convertStringSetToInterface(respTypes),
+		"grant_types":               convertStringSetToInterface(grantTypes),
+		"post_logout_redirect_uris": convertStringSetToInterface(app.Settings.OauthClient.PostLogoutRedirectUris),
+	}
+	err = setNonPrimitives(d, aggMap)
+	if err != nil {
+		return diag.Errorf("failed to set OAuth application properties: %v", err)
+	}
+	return nil
+}

--- a/okta/data_source_okta_app_oauth_test.go
+++ b/okta/data_source_okta_app_oauth_test.go
@@ -1,0 +1,54 @@
+package okta
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccOktaDataSourceAppOauth_read(t *testing.T) {
+	ri := acctest.RandInt()
+	mgr := newFixtureManager(appOAuth)
+	config := mgr.GetFixtures("datasource.tf", ri, t)
+	appCreate := buildTestAppOauth(ri)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProvidersFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: appCreate,
+			},
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("data.okta_app_oauth.test", "client_id"),
+					resource.TestCheckResourceAttrSet("data.okta_app_oauth.test", "grant_types.#"),
+					resource.TestCheckResourceAttrSet("data.okta_app_oauth.test", "redirect_uris.#"),
+					resource.TestCheckResourceAttrSet("data.okta_app_oauth.test", "type"),
+					resource.TestCheckResourceAttr("data.okta_app_oauth.test", "label", fmt.Sprintf("testAcc_%d", ri)),
+					resource.TestCheckResourceAttr("data.okta_app_oauth.test_label", "label", fmt.Sprintf("testAcc_%d", ri)),
+					resource.TestCheckResourceAttr("data.okta_app_oauth.test", "status", statusActive),
+					resource.TestCheckResourceAttr("data.okta_app_oauth.test_label", "status", statusActive),
+				),
+			},
+		},
+	})
+}
+
+func buildTestAppOauth(d int) string {
+	return fmt.Sprintf(`
+resource "okta_app_oauth" "test" {
+  label                      = "testAcc_%d"
+  type                       = "web"
+  grant_types                = ["authorization_code"]
+  redirect_uris              = ["http://d.com/"]
+  response_types             = ["code"]
+  client_basic_secret        = "something_from_somewhere"
+  client_id                  = "something_from_somewhere"
+  token_endpoint_auth_method = "client_secret_basic"
+  consent_method             = "TRUSTED"
+}`, d)
+}

--- a/okta/provider.go
+++ b/okta/provider.go
@@ -224,6 +224,7 @@ func Provider() *schema.Provider {
 		DataSourcesMap: map[string]*schema.Resource{
 			"okta_app":                         dataSourceApp(),
 			"okta_app_saml":                    dataSourceAppSaml(),
+			appOAuth:                           dataSourceAppOauth(),
 			"okta_app_metadata_saml":           dataSourceAppMetadataSaml(),
 			"okta_default_policies":            deprecatedPolicies,
 			"okta_default_policy":              dataSourceDefaultPolicies(),

